[Core][Model] PrithviMAE Enablement on vLLM v1 engine

christian-pinto opened the pull request:
This is a PR to be used for comparison with #20072 where support for attention free models was done via overloading the `KVCacheManager`.

christian-pinto committed:
Better support for skip_tokenizer_init=True

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Support passing raw multimodal data to model

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Support for attention free models in V1

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Last few changes after rebasing to latest branch version

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
latest changes to align with the original branch

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Latest changes to aadpt to upstream master

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Some reformatting to make the pre-commit hooks succeed

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Few more changes to solve some other pre-commit hooks failures

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Some style changes

- Improved formatting around
- made is_pooling_model a @property in ModelConfig

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Simple code refactoring
- Remove unused functions
- merged functions not called anywhere else

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

heheda12345 commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -438,29 +446,29 @@ def schedule(self) -> SchedulerOutput:
-                # needed for this request.
-                if self.connector is not None:
-                    self.connector.update_state_after_alloc(
+                if not self.vllm_config.model_config.is_attention_free:
```
Why don't we need these lines in #20072 ?

christian-pinto replied:
Just a wild indentation!

heheda12345 commented on vllm/v1/engine/core.py:
```diff
@@ -134,21 +134,33 @@ def _initialize_kv_caches(
-            zip(kv_cache_specs, available_gpu_memory)
-        ]
+        #TODO: CP start from here
+        if vllm_config.model_config.is_attention_free:
```
I prefer to handle attention free model here https://github.com/vllm-project/vllm/blob/3eb4ad53f3e4ed0d3224c37efd08b4a1a42076fc/vllm/v1/core/kv_cache_utils.py#L927 . Would be great if it can be achieved by another branch in addtion to `is_kv_cache_type_uniform` and `is_kv_cache_page_size_uniform ` and pls tell me if further modfications are needed.

heheda12345 commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -295,7 +295,9 @@ def add_request(
         self.num_tokens_no_spec[req_index] = request.num_tokens
 
         self.num_computed_tokens_cpu[req_index] = request.num_computed_tokens
-        self.block_table.add_row(request.block_ids, req_index)
```
I think this line does nothing if there is 0 kv cache groups.

heheda12345 commented on vllm/v1/core/block_pool.py:
```diff
@@ -36,7 +36,8 @@ def __init__(
     ):
-        assert isinstance(num_gpu_blocks, int) and num_gpu_blocks > 0
+        # num_gpu_blocks can be 0 for attention free models
+        assert isinstance(num_gpu_blocks, int)
```
I think we can always have at least 1 gpu block so that we don't need to handle null block.

heheda12345 commented on vllm/v1/core/kv_cache_utils.py:
```diff
@@ -200,14 +200,18 @@ class FreeKVCacheBlockQueue:
-                blocks[i].prev_free_block = blocks[i - 1]
-            if i < self.num_free_blocks - 1:
-                blocks[i].next_free_block = blocks[i + 1]
+        # This is 0 in attention free models
```
These modifications are not needed if with at least 1 gpu block

heheda12345 commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -92,7 +92,9 @@ def __init__(
-        assert num_gpu_blocks is not None and num_gpu_blocks > 0
+
+        # num_gpu_blocks can be zero for attention free models
+        assert num_gpu_blocks is not None
```
Not needed if with at least 1 gpu block

heheda12345 commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -246,6 +248,12 @@ def schedule(self) -> SchedulerOutput:
             while True:
+                # This model is attention free and we do not need to allocate KVCache blocks
+                # for serving requests.
+                if self.vllm_config.model_config.is_attention_free:
```
Why do you need this change? I think allocate_slots should always succeed,

heheda12345 commented on vllm/v1/core/kv_cache_coordinator.py:
```diff
@@ -219,7 +219,9 @@ def __init__(self, kv_cache_config: KVCacheConfig, max_model_len: int,
-        self.verify_and_split_kv_cache_groups()
+        # attention free models are initialized with 0 kv_cache_groups
+        if len(self.kv_cache_config.kv_cache_groups) > 0:
+            self.verify_and_split_kv_cache_groups()
```
I'm comfortable with adding another coordinator for 0 kv cache groups and re-implement find_longest_cache_hit for it.

nopperl replied:
@christian-pinto I have introduced a `KVCacheCoordinatorNoPrefixCache` in this PR (#20661 ). I think it should handle your case as well. Could you give it a try?

christian-pinto replied:
Hey @nopperl thanks for that. Your approach solves my issue too.

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -327,6 +329,11 @@ def _may_reorder_batch(self, scheduler_output: "SchedulerOutput") -> None:
         """
+        
+        # nothing to be reordered when the mdoel is attention free
+        if self.model_config.is_attention_free:
```
I'm OK with this change temporarily. I'll refactor this function soon to handle both attention-free case and many other unsupported cases.

heheda12345 reviewed (commented):
I checked the code in `block_pool.py`, `kv_cache_coordinator.py`, `kv_cache_manager.py`, `kv_cache_utils.py`, `core.py` `gpu_input_batch.py` and feel that it is possible to implement 0 kv cache config cleanly. I'm not sure whether the changes in scheduler.py is required.

It's quite hard for me to find all code related to kv cache manager. Can you help to put them into a separate PR if you want to continue with 0 kv cache group option?

heheda12345 reviewed (commented):
Yeah I like the current implementation of 0 kv cache group. Can you make a new PR for that?

christian-pinto committed:
Rebased to master

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Ensure pre-commit checks succeed

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Few more style edits

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@nopperl @DarkLight1337 @maxdebayser I have rebased this PR after my other changes in #20811 regarding KV cache management got merged, please take a look.

Also, I have added a test script for this model that verifies the model can be invoked but does not compare with inferencing via transformers. However, when I try to run the test on my environment it gets stuck on initializing the engine.

https://github.com/vllm-project/vllm/blob/d0dc4cfca48c2734da18ec42d6bba1346cbfc400/vllm/v1/engine/utils.py#L478

Any idea why this could be happening?

christian-pinto committed:
Skip tokenizer init in async LLM engine

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 reviewed (commented):
Can you merge in the latest changes?

christian-pinto committed:
Merge branch 'main' into prithvi_v1_embeddings_zero_kv_cache_group

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/model_executor/models/interfaces.py:
```diff
@@ -136,6 +136,48 @@ def supports_multimodal(
+    model: Union[type[object], object]
+) -> Union[TypeIs[type[SupportsMultiModalWithRawInput]],
+           TypeIs[SupportsMultiModalWithRawInput]]:
+    if isinstance(model, type):
+        return isinstance(model, _SupportsMultiModalWithRawInput)
+
+    return isinstance(model, SupportsMultiModalWithRawInput)
```
Change this to `getattr`

DarkLight1337 commented on vllm/model_executor/models/prithvi_geospatial_mae.py:
```diff
@@ -99,24 +103,36 @@ def apply(
 
         for k, v in mm_data.items():
             mm_kwargs[k] = v
+        mm_place_holders = {"image": [PlaceholderRange(offset=0, length=0)]}
```
```suggestion
        mm_placeholders = {"image": [PlaceholderRange(offset=0, length=0)]}
```

Typo

DarkLight1337 commented on vllm/model_executor/models/prithvi_geospatial_mae.py:
```diff
@@ -99,24 +103,36 @@ def apply(
+
+class PrithviGeoSpatialMAE(nn.Module, IsAttentionFree,
+                           SupportsMultiModalWithRawInput):
+    """ Prithvi Masked Autoencoder"""
```
```suggestion
    """Prithvi Masked Autoencoder"""
```

DarkLight1337 commented on vllm/model_executor/models/prithvi_geospatial_mae.py:
```diff
@@ -185,6 +200,15 @@ def _parse_and_validate_multimodal_data(
+        # We do not really use any input tokens and therefore no embeddings
+        # to be calculated. However, due to the mandatory token ids in
+        # the input prompt we pass one token and the size of the dummy
+        #  embedding tensors must reflect that.
```
```suggestion
        # embedding tensors must reflect that.
```

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -568,6 +573,38 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
         # Refresh batch metadata with any pending updates.
         self.input_batch.refresh_metadata()
 
+    def _maybe_add_multimodal_kwargs(
```
I would rename this to avoid being confused with the multimodal inputs that are used in `_execute_mm_encoder`

christian-pinto committed:
Changes after review

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/config.py:
```diff
@@ -642,6 +642,8 @@ def __post_init__(self) -> None:
         self.max_model_len = self.get_and_verify_max_len(self.max_model_len)
         self.multimodal_config = self._init_multimodal_config()
+        self.model_supports_multimodal_raw_input = (
+            self._init_model_supports_multimodal_raw_input())
```
Is it really necessary to make this a separate method?

christian-pinto replied:
I took inspiration from other properties of the model config that go through the registry to figure out whether the model has a specific property.

If there is another preferred way of doing this I am happy to follow it.

DarkLight1337 replied:
I think you can just inline the call

DarkLight1337 commented on vllm/model_executor/models/interfaces.py:
```diff
@@ -136,6 +136,45 @@ def supports_multimodal(
+    """
+
+
+@runtime_checkable
+class _SupportsMultiModalWithRawInput(Protocol):
+    supports_multimodal_raw_input: ClassVar[Literal[True]]
```
Can remove this

DarkLight1337 commented on vllm/v1/engine/processor.py:
```diff
@@ -380,7 +380,9 @@ def _validate_model_input(
         model_config = self.model_config
-        tokenizer = self.tokenizer.get_lora_tokenizer(lora_request)
+
+        tokenizer = (None if model_config.skip_tokenizer_init else
+                     self.tokenizer.get_lora_tokenizer(lora_request))
```
I would merge this with L394

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1347,15 +1381,20 @@ def execute_model(
             # as input to the multimodal model, even when the input is text.
             input_ids = self.input_ids[:num_scheduled_tokens]
+
+            self._maybe_update_model_kwargs(model_kwargs=model_kwargs,
```
I prefer setting `model_kwargs = {}` explicitly in the `else` branch and generate new `model_kwargs` here

DarkLight1337 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2008,7 +2051,10 @@ def _dummy_run(
         with self.maybe_dummy_run_with_lora(self.lora_config,
                                             num_scheduled_tokens):
             model = self.model
+            model_kwargs: dict[str, Any] = {}
```
Same as above

christian-pinto committed:
Further review round

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Further edits

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
One last if condition to be flipped

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
pre-commit hooks failing

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Added tokenizer check in AsyncLLM.get_tokenizer()

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented on vllm/model_executor/models/prithvi_geospatial_mae.py:
```diff
@@ -99,23 +101,34 @@ def apply(
             mm_kwargs[k] = v
+        mm_placeholders = {"image": [PlaceholderRange(offset=0, length=0)]}
+
+        multimodal_kwargs_items = [
+            MultiModalKwargsItem.from_elems([
+                MultiModalFieldElem(modality="image",
+                                    key=key,
+                                    data=data,
+                                    field=MultiModalSharedField(1))
+                for key, data in mm_kwargs.items()
+            ])
+        ]
```
Yes, with this set of changes it does.

DarkLight1337 replied:
Hmm, we normally expect to have one `PlaceholderRange` per image. But now it looks like you only have one for the whole batch.

christian-pinto replied:
OK let me have a final look at this on monday as there could be a double source of batching:
- The model receives a batch of image patches in one request
- vLLM batches requests together (regardless of the number of image patches per request)

christian-pinto replied:
If each request comes with multiple image patches, should I then have one PlaceholderRange for each of them? 
The model would anyway treat them as one batch and perform an inference on all of them at once.

DarkLight1337 replied:
In the more general case, the leading dimension is treated as the number of input images if you use `MultiModalFieldConfig.batched`.

christian-pinto replied:
Ok, thanks

christian-pinto replied:
I did a few experiments here. For me the best type of field is indeed the Shared one with bs=1. This is because I want that all the multimodal data that is passed in input with the "prompt" to be considered as one single element. 
shape of the multimodal data with the prompt is: (num_channels, width, height). I am assuming one patch per prompt.
When batching multiple requests together, vLLM will add another dimension and the shape becomes (vllm_batch_size, num_channels, width, height).

At this stage I am not concerned with the need for submitting multiple patches with the same request since I can still get the batching done for free by vLLM. Therefore I would keep the current implementation as is.
Also, one placeholder is also fine because I do consider the whole input as one Element.

christian-pinto replied:
In the near future I will have to work out how to deal with both multiple input patches and vLLM batching at the same time . This is because patches and batches can be "batched" together with this model.

The model would receive and input with the below shape

(vllm_batch_size, num_patches, num_channels, width, height).

`vllm_batch_size` and `num_patches` need to be merged together as batching more requests translates for the model into processing more patches in parallel.  

And I believe this could probably be achieved with  `einops.rearrange` in the model forward function to merge the two dimensions, perform the inference and split them again before returning the output.
I would prefer to implement this in a following PR only with changes to the model only..

christian-pinto committed:
Merge branch 'main' into prithvi_v1_embeddings_zero_kv_cache_group

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 reviewed (approved):
LGTM, thanks for working on it!

DarkLight1337 commented:
Can you try to increase the timeout and see how long it needs to start up the model?

christian-pinto committed:
Added PrithviMAE test requirements

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Testing test deadlock fix

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Merge branch 'main' into prithvi_v1_embeddings_zero_kv_cache_group

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
I have committed both the test fix and one change in the tests requirement. If all is good tomorrow we will publish a new version of the `terratorch` library and I will re-update the requirements file. We should be good then.

christian-pinto committed:
Pre-commit of course

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fix GPUModelRunner._may_reorder_batch to support Mamba models

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented:
I think they are using L4s (24 GB VRAM)

christian-pinto committed:
Updated iprithvi geospatial MAE test to avoid OOM during warmup

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
> I think they are using L4s (24 GB VRAM)

It was OOMing during the warmup run as it tries with 256 requests in parallel. That's fixed too. I'll push a new and hopefully final version soon

christian-pinto committed:
Updated requirements for PrithviGeospatialMAE tests

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@DarkLight1337 this should be it now. If all tests succeed as I expect, I am happy for this version to get merged. Thanks

christian-pinto commented:
@DarkLight1337 If needed for merging, could you please restart the `engine-test`?

It failed because of a hugging face HTTP timeout

```
[2025-07-23T15:39:39Z] ERROR tokenization/test_detokenize.py::test_decode_streaming[True-True-False-mistralai/Pixtral-12B-2409-False-Hello here, this is a simple test] - huggingface_hub.errors.HfHubHTTPError: 504 Server Error: Gateway Timeout for url: https://huggingface.co/api/models/mistralai/Pixtral-12B-2409/tree/main?recursive=True&expand=False (Request ID: Root=1-6880ff4b-277e30b23bcda172710b9025;a986c640-f5b2-4b6e-9874-475e1a90e0cc)
```

Many thanks

christian-pinto commented:
@DarkLight1337 my bad, I have mistakenly closed the PR and got automerge disabled.

Sorry about that.
